### PR TITLE
Traited Render System and Reload tracking

### PR DESCRIPTION
**Pull request by QuantumEntangledAndy**
Depends on #170

---

This is a bit of a bigger PR that adds the following:

- Use traits for binding to the pipeline. This means we can bind like this:
    ```rust
    vec![Binding::texture_cube("CubeMap", Stage::Fragment, cubemap)],
    ```
    instead of this:
    ```rust
    vec![Binding::texture_cube("CubeMap", Stage::Fragment, &cubemap.uniform)],
    ```
    - This is achieved with the traits in `dotrix_core::providers::*`
- Add a reload system to signal when an asset needs to be rebound to the pipeline.
   There are two kinds of reload, `ReloadKind::Update` and `ReloadKind::Reload` the first is when data was updated but not the length of the buffer or dimensions of the texture, the second is when these are changed and when a rebind is required.
    - This rebind is done automatically during a `bind` of the pipeline. Which is now a no-op if no rebind has been signalled as required. We can therefore just bind everyframe and let the underlying system handle checking for the need to rebinding.
- Also a `dotrix_derive` that adds derive macros to help add reloadable and other traits

Comments are requested as this is kinda far reaching in its effects and we can maybe discuss alternative methods and techniques

### CODE REVIEW

**Comment by Lowentwickler** [comment]
Wow! Going to check it out!

**Comment by QuantumEntangledAndy** [comment]
I have a few ideas on improvements already but will have to see if they pan out the way I think they will

File: dotrix_core/src/assets/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+/// Buffer asset
+#[derive(Reloadable, BufferProvider)]
+pub struct Buffer {
+    /// The length of the data that the buffer was created with
+    pub data_len: usize,
+    /// The underlying gpu buffer
+    pub buffer: GpuBuffer,
+    /// Flagged on change
+    pub reload_state: ReloadState,
+}
```
Comment (Lowentwickler) [inline dotrix_core/src/assets/buffer.rs:19]: [Question]
What about using GpuBuffer itself as an asset? Why we need a wrapper?

File: dotrix_core/src/reloadable.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+}
+
+/// Describes an object that has gpu represntable data
+/// that may update and needs periodic (re-)loading
+///
+/// A function that needs to for check changes since an Instance
+/// should call [`changed_since`]
+pub trait Reloadable {
```
Comment (Lowentwickler) [inline dotrix_core/src/reloadable.rs:48]: [question]
Could you please describe a use case when there will be just one `Reloadable` or `ReloadableMut` trait implemented?

Comment (QuantumEntangledAndy) [reply dotrix_core/src/reloadable.rs:48]: Oh, well when you are rendering you only have access to `&` reference so you can only read the reload state. When you are editing you have an `&mut` reference so you can flag it as dirty. You cannot have a function that takes an `&mut self` in a trait impl of `& MyStruct` so you have to seperate them.

File: dotrix_core/src/assets/buffer.rs (hunk 1 above)
Comment (QuantumEntangledAndy) [reply dotrix_core/src/assets/buffer.rs:19]: Was debating this but went with this for kind of seperation of concern. This handles the high level representation, while the Renderer one the low level. Originally this struct behaved a bit like `assets::Texture` in that it held a copy of the data in a `Vec` but there was no real use case for it over just direct loading.

It also made it obvious to `impl BufferProvider` which allows the abstraction of having any struct provide a reference to a `GpuBuffer` for the Pipieline, although I suppose `GpuBuffer` could `impl BufferProvider` and return `&self`

File: dotrix_core/src/reloadable.rs (hunk 2 above)
Comment (QuantumEntangledAndy) [reply dotrix_core/src/reloadable.rs:48]: I will see if I can explain this.

At bind you have:

```rust
Binding::Texture("label", Stage::All, Box<& Texture>(asset)
```

Notice that it is a `&Texture` which comes from `assets.get<Texture>(id)`.

The trait `Reloadable` cannot have mut assess to this, so the trait impl must not have any `&mut self`  functions if I want to use this trait.

**Pull request closed without merging**